### PR TITLE
feat: add multiple features and fix

### LINKED ISSUE #37: Add label-name list in exported output
We currently use promlinter(version v0.0.2) to export our project's metrics to documents, it works well, but it seems there is no `--add-label`, and I think label is a important part to understand the metrics.

### PR DESCRIPTION
**Pull request by coanor**
- Enable list metrics in vendor packages(`--with-vendor`)
    - Why? Some vendor packages also export metrics, we need to export them for full metric documents.
- List labels of metrics, if no labels, output `N/A`(#37)
- Fix panic on `--add-help` if metric got no help info, and also if no help output `N/A`
- Out JSON with pretty indents
- Output markdown table(#38)
- Avoid duplicated metric name(#40)

Close #37/#38/#40.

### CODE REVIEW

**Comment by coanor** [comment]
The output example seems like this:

```shell
TYPE                NAME                    LABELS       HELP
COUNTER             http_api_signed_total   api,method   API signature count
COUNTER             http_api_dropped_total  api,method   API request dropped when sinker rule match failed
```

Labels are `,` splited string. If there're const labels, the label displayed as:

- `some_const_label=some_value`: The const label got a string literal value
- `some_const_label=?`: While the const label value are not string literal(such as from `fmt.Sprintf()` or `someModule.SomeValue`)

And the JSON format like this:

```json
[
  {
    "Name": "http_api_dropped_total",
    "Help": "API request dropped when sinker rule match failed",
    "Type": "COUNTER",
    "Filename": "/path/to/metrics.go",
    "Labels": [
      "api",
      "method"
    ],
    "Line": 105,
    "Column": 3
  },
  {
    "Name": "http_api_signed_total",
    "Help": "API signature count",
    "Type": "COUNTER",
    "Filename": "/path/to/metrics.go",
    "Labels": [
      "api",
      "method"
    ],
    "Line": 118,
    "Column": 3
  }
]
```

The markdown table seems like this:

```markdown
| TYPE    | NAME                     | LABELS       | HELP                                              |
| ---     | ---                      | ---          | ---                                               |
| COUNTER | `http_api_signed_total`  | `api,method` | API signature count                               |
| COUNTER | `http_api_dropped_total` | `api,method` | API request dropped when sinker rule match failed |
```

**Comment by yeya24** [comment]
Thanks for the good work! @coanor Are you able to add some unit tests about the labels feature?

**Comment by coanor** [comment]
> Thanks for the good work! @coanor Are you able to add some unit tests about the labels feature?

I find the PR can't find labels within `NewDesc`, I need some time to refactor current `parseXXX` functions. And current PR closed, I'll reopen it after `variableLabels` and `constLabels` able to parse.

**Comment by coanor** [comment]
> Thanks for the good work! @coanor Are you able to add some unit tests about the labels feature?

I've add label discovery within `NewDesc` and basic test case, but it seems that the original case within *promlinter_test.go* failed, there missing a issue in current MR, currently I have no idea about what happend.

Will you give me some hint about why the case failed?

**Comment by yeya24** [comment]
It seems like the error case below was missing.

```
	if issues[2].Metric != "test_metric_total" && issues[2].Text != `no help text` {
		t.Fatal()
	}
```

**Comment by coanor** [comment]
> It seems like the error case case below was missing.
> 
> ```
> 	if issues[2].Metric != "test_metric_total" && issues[2].Text != `no help text` {
> 		t.Fatal()
> 	}
> ```

I've refactor and fixed the testings within *promlinter_test.go*, It seems that the issue's order not the key concerning.

File: promlinter.go [hunk 1]
```diff
@@ -254,19 +291,44 @@ func (v *visitor) parseCallerExpr(call *ast.CallExpr) ast.Visitor {
...
 		return v
 	}
+
 	currentMetric := dto.MetricFamily{
 		Type: &metricType,
-		Help: help,
+	}
+
+	if !opts.helpSet {
+		currentMetric.Help = nil
```
Comment (yeya24) [inline promlinter.go:322]: We don't need this condition. We cannot set String to `nil` I guess.
Line 327 should be enough to check if help is set

File: promlinter.go [hunk 2]
```diff
@@ -395,17 +483,50 @@ func (v *visitor) parseOptsExpr(n ast.Node) (*opt, *string) {
...
-	var help *string
+
 	for _, elt := range stmt.Elts {
+
+		// labels
+		label, ok := elt.(*ast.BasicLit)
+		if ok {
+			metricOption.labels = append(metricOption.labels, strings.Trim(label.Value, `"`))
+			continue
+			// go down
```
Comment (yeya24) [inline promlinter.go]: Do we need this comment?

**Comment by yeya24** [comment]
> I've refactor and fixed the testings within promlinter_test.go, It seems that the issue's order not the key concerning.

I am ok to refactor the tests but I think it is still good to return ordered (at least order can be determined) issues from `RunLint()`.

File: promlinter.go (hunk 1 above)
Comment (coanor) [reply promlinter.go:322]: `currentMetric.Help` is type `*string`, we just assign nil to mean "There is no help available". If we set a point to empty string(`""`, here is line 327's `opts.help`, it's default value is `""`), some exists test cases will fail(*empty string still a valid help, but not make sense.*).

File: promlinter.go (hunk 2 above)
Comment (coanor) [reply promlinter.go]: No, this may a noise comment, just removed.

**Comment by coanor** [comment]
> > I've refactor and fixed the testings within promlinter_test.go, It seems that the issue's order not the key concerning.
> 
> I am ok to refactor the tests but I think it is still good to return ordered (at least order can be determined) issues from `RunLint()`.

About the sort, I think we do not need to sort the result issue, because they are sorted already according to the AST traverse(And I removed the trailing `sort` on issues(fd4631db6cd98966c40b2d6b2d8740644d30d55f)).

For example, the raw issue list is(from *testdata/testdata.go*)

```
test_metric_name: ./testdata/testdata.go:23:3
test_metric_total: ./testdata/testdata.go:32:3
foo: ./testdata/testdata.go:39:30
foo_bar_total: ./testdata/testdata.go:78:8
kube_test_metric_count: ./testdata/testdata.go:92:8
test_histogram_duration_seconds: ./testdata/testdata.go:95:27
kube_daemonset_labels: ./testdata/testdata.go:119:4
```

After sort(based on `pos.String()`), the ording changed(and is not the AST order):

```
kube_daemonset_labels: ./testdata/testdata.go:119:4
test_metric_name: ./testdata/testdata.go:23:3
test_metric_total: ./testdata/testdata.go:32:3
foo: ./testdata/testdata.go:39:30
foo_bar_total: ./testdata/testdata.go:78:8
kube_test_metric_count: ./testdata/testdata.go:92:8
test_histogram_duration_seconds: ./testdata/testdata.go:95:27
```

Because string `a.go:123 < a.go:99`, and that *not* the order we need：

```go
import (
  T "testing"

  "github.com/stretchr/testify/assert"
)

func TestStringOrder(t *T.T) {
	assert.True(t, "a.go:123" < "a.go:99")  // Pass
}
```

**Comment by yeya24** [comment]
Sorry for taking it for so long. I am going to merge it. Great work! Thanks

**Review by yeya24 — APPROVED** [review]

**Pull request merged**